tooling/templatize/pkg/pipeline/types.go

```diff
@@ -14,25 +16,170 @@ type Pipeline struct {
 }
 
 type ResourceGroup struct {
-	Name         string  `yaml:"name"`
-	Subscription string  `yaml:"subscription"`
-	AKSCluster   string  `yaml:"aksCluster,omitempty"`
-	Steps        []*Step `yaml:"steps"`
+	Name         string `yaml:"name"`
+	Subscription string `yaml:"subscription"`
+	AKSCluster   string `yaml:"aksCluster,omitempty"`
+	Steps        []Step `yaml:"steps"`
+}
+
+func (rg *ResourceGroup) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	rawRg := &struct {
+		Name         string    `yaml:"name"`
+		Subscription string    `yaml:"subscription"`
+		AKSCluster   string    `yaml:"aksCluster,omitempty"`
+		Steps        []rawStep `yaml:"steps"`
+	}{}
+	if err := unmarshal(&rawRg); err != nil {
+		return err
+	}
+	rg.Name = rawRg.Name
+	rg.Subscription = rawRg.Subscription
+	rg.AKSCluster = rawRg.AKSCluster
+	rg.Steps = make([]Step, len(rawRg.Steps))
+	for i, rawStep := range rawRg.Steps {
+		switch rawStep.meta.Action {
+		case "Shell":
+			rg.Steps[i] = &ShellStep{}
+		case "ARM":
+			rg.Steps[i] = &ARMStep{}
+		default:
+			return fmt.Errorf("unknown action type %s", rawStep.meta.Action)
+		}
+		err := rawStep.unmarshal(rg.Steps[i])
```

Comment:
don't you have to call UnmarshalYAML?

Reply:
`UnmarshalYAML` this is a "magic" function the golang yaml library calls